File: src/main/java/com/enderryno/nuclearcraft/classes/CustomBlock.java
package com.enderryno.nuclearcraft.classes;


import com.enderryno.nuclearcraft.NuclearCraft;
import com.enderryno.nuclearcraft.database.definitions.BlockTable;
import com.enderryno.nuclearcraft.enums.BlockBehaviour;
import com.enderryno.nuclearcraft.interfaces.PluginBlock;
import com.enderryno.nuclearcraft.interfaces.PluginItem;
import com.enderryno.nuclearcraft.enums.ItemBehaviour;
import com.enderryno.nuclearcraft.services.BlockRegister;
import com.jeff_media.customblockdata.CustomBlockData;
import org.bukkit.Location;
import org.bukkit.NamespacedKey;
import org.bukkit.block.Block;
import org.bukkit.block.BlockState;
import org.bukkit.metadata.FixedMetadataValue;
import org.bukkit.metadata.MetadataValue;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;

import java.sql.ResultSet;


/**
 * This class instantiates a custom item by its id
 * <p>
 * Since it's unsafe to explicitly extend ItemStack class,
 * this class has a getter for both the ItemStack instance and this plugin item class.
 */
public class CustomBlock implements PluginBlock {
    private String id;
    private String minecraftId;
    private BlockBehaviour behaviour;

    int customModelId;

    @Override
    public PluginBlock setID(String id) {
        this.id = id;
        return this;
    }

    @Override
    public String getID() {
        return this.id;
    }

    @Override
    public PluginBlock setCustomModelId(int customModelId) {
        this.customModelId = customModelId;
        return this;
    }

    @Override
    public int getCustomModelId() {
        return this.customModelId;
    }

    @Override
    public PluginBlock setMinecraftId(String minecraftId) {
        this.minecraftId = minecraftId;
        return this;
    }

    @Override
    public String getMinecraftId() {
        return this.minecraftId;
    }


    @Override
    public void placeBlock(PluginItem item, Location location) {
        // Get block at location
        Block block = location.getWorld().getBlockAt(location);
        // Save metadata
        PersistentDataContainer blockData = new CustomBlockData(block, NuclearCraft.instance);

        NamespacedKey blockIdKey = new NamespacedKey(NuclearCraft.instance, "custom-block-id");
        NamespacedKey itemIdKey = new NamespacedKey(NuclearCraft.instance, "item-id");

        blockData.set(blockIdKey, PersistentDataType.STRING, this.getID());
        blockData.set(itemIdKey, PersistentDataType.STRING, item.getID());

        // Print in chat for debugging
        NuclearCraft.instance.getLogger().info("Placed block " + blockData.get(blockIdKey, PersistentDataType.STRING));
    }

    @Override
    public PluginBlock getBlock(Location location) {
        // Try to get block metadata
        Block block = location.getWorld().getBlockAt(location);
        NamespacedKey blockIdKey = new NamespacedKey(NuclearCraft.instance, "custom-block-id");
        PersistentDataContainer blockData = new CustomBlockData(block, NuclearCraft.instance);
        String customBlockId = blockData.get(blockIdKey, PersistentDataType.STRING);
        try {
            BlockRegister.getRegisteredBlocks().get(customBlockId);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
        return null;
    }

    @Override
    public void removeBlock(Location location) {
        // Get block at location
        Block block = location.getWorld().getBlockAt(location);
        PersistentDataContainer blockData = new CustomBlockData(block, NuclearCraft.instance);

        NamespacedKey blockIdKey = new NamespacedKey(NuclearCraft.instance, "custom-block-id");
        NamespacedKey itemIdKey = new NamespacedKey(NuclearCraft.instance, "item-id");

        blockData.remove(blockIdKey);
        blockData.remove(itemIdKey);
    }

    @Override
    public PluginBlock setBehaviour(BlockBehaviour behaviour) {
        this.behaviour = behaviour;
        return this;
    }

    @Override
    public BlockBehaviour getBehaviour() {
        return this.behaviour;
    }
}

File: src/main/java/com/enderryno/nuclearcraft/classes/CustomBlock.kt
package com.enderryno.nuclearcraft.classes

import com.enderryno.nuclearcraft.NuclearCraft
import com.enderryno.nuclearcraft.enums.BlockBehaviour
import com.enderryno.nuclearcraft.interfaces.PluginBlock
import com.enderryno.nuclearcraft.interfaces.PluginItem
import com.enderryno.nuclearcraft.services.BlockRegister
import com.jeff_media.customblockdata.CustomBlockData
import org.bukkit.Location
import org.bukkit.NamespacedKey
import org.bukkit.persistence.PersistentDataContainer
import org.bukkit.persistence.PersistentDataType

/**
 * This class instantiates a custom item by its id
 *
 */
class CustomBlock(
    override val id: String,
    override val customModelId: Int,
    override val minecraftId: String,
    override val behaviour: BlockBehaviour
) : PluginBlock{

    override fun placeBlock(item: PluginItem, location: Location) {
        // Get block at location
        val block = location.world.getBlockAt(location)
        // Save metadata
        val blockData: PersistentDataContainer = CustomBlockData(block, NuclearCraft.instance!!)
        val blockIdKey = NamespacedKey(NuclearCraft.instance!!, "custom-block-id")
        val itemIdKey = NamespacedKey(NuclearCraft.instance!!, "item-id")
        blockData.set(blockIdKey, PersistentDataType.STRING, id!!)
        blockData.set(itemIdKey, PersistentDataType.STRING, item?.id!!)

        // Print in chat for debugging
        NuclearCraft.instance!!.getLogger().info("Placed block " + blockData.get(blockIdKey, PersistentDataType.STRING))
    }

    override fun removeBlock(location: Location) {
        // Get block at location
        val block = location.world.getBlockAt(location)
        val blockData: PersistentDataContainer = CustomBlockData(block, NuclearCraft.instance!!)
        val blockIdKey = NamespacedKey(NuclearCraft.instance!!, "custom-block-id")
        val itemIdKey = NamespacedKey(NuclearCraft.instance!!, "item-id")
        blockData.remove(blockIdKey)
        blockData.remove(itemIdKey)
    }

}
